LibreNMS/Polling/ConnectivityHelper.php

```diff
@@ -0,0 +1,216 @@
+use Symfony\Component\Process\Process;
+
+class ConnectivityHelper
+{
+    /**
+     * @var \App\Models\Device
+     */
+    private $device;
+    /**
+     * @var bool
+     */
+    private $saveMetrics = false;
+    /**
+     * @var string
+     */
+    private $family;
+    /**
+     * @var string
+     */
+    private $target;
+
+    public function __construct(Device $device)
+    {
+        $this->device = $device;
+        $this->target = $device->overwrite_ip ?: $device->hostname;
+    }
+
+    /**
+     * After pinging the device, save metrics about the ping response
+     */
+    public function saveMetrics(): void
+    {
+        $this->saveMetrics = true;
+    }
+
+    /**
+     * Check if the device is up.
+     * Save availability and ping data if enabled with savePingPerf()
+     */
+    public function isUp(): bool
+    {
+        $previous = $this->device->status;
+        $ping_response = $this->isPingable();
+
+        // calculate device status
+        if ($ping_response->success()) {
+            if (! $this->canSnmp() || $this->isSNMPable()) {
+                // up
+                $this->device->status = true;
+                $this->device->status_reason = '';
+            } else {
+                // snmp down
+                $this->device->status = false;
+                $this->device->status_reason = 'snmp';
+            }
+        } else {
+            // icmp down
+            $this->device->status = false;
+            $this->device->status_reason = 'icmp';
+        }
+
+        if ($this->saveMetrics) {
+            if ($this->canPing()) {
+                $this->savePingStats($ping_response);
+            }
+            $this->updateAvailability($previous, $this->device->status);
+
+            $this->device->save(); // confirm device is saved
+        }
+
+        return $this->device->status;
+    }
+
+    /**
+     * Check if the device responds to ICMP echo requests ("pings").
+     */
+    public function isPingable(): FpingResponse
+    {
+        if (! $this->canPing()) {
+            return FpingResponse::artificialUp();
+        }
+
+        $status = app()->make(Fping::class)->ping(
+            $this->target,
+            Config::get('fping_options.count', 3),
+            Config::get('fping_options.interval', 500),
+            Config::get('fping_options.timeout', 500),
+            $this->ipFamily()
+        );
+
+        if ($status->duplicates > 0) {
+            Log::event('Duplicate ICMP response detected! This could indicate a network issue.', $this->device, 'icmp', 4);
+            $status->exit_code = 0;   // when duplicate is detected fping returns 1. The device is up, but there is another issue. Clue admins in with above event.
+        }
+
+        return $status;
+    }
+
+    public function isSNMPable(): bool
+    {
+        $response = NetSnmp::device($this->device)->get('SNMPv2-MIB::sysObjectID.0');
+
+        return $response->getExitCode() === 0 || $response->isValid();
```

Comment:
|| vs &&,

I'm guessing not all devices responds to sysobjectid and therefor wouldn't be "valid"?
Maybe deserves a comment?
There is snmpping also but it just seems to fetch some other oids

Reply:
This is correct matches with previous behavior.  I was very careful about this part as I messed with this before and had to revert. :D
Actually, old code sent one snmpget, check the return code was 0 if it was great.
If not, it sent another request and checked the output.